compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -931,6 +931,7 @@ pub enum ProjectionElem<V, T> {
     ///
     /// The `x[i]` is turned into a `Deref` followed by an `Index`, not just an `Index`. The same
     /// thing is true of the `ConstantIndex` and `Subslice` projections below.
+    #[skip_traversal]
```

Comment:
can you move this attribute to the generic parameter itself?

i.e. parameters with `skip_traversal` get a `V: SkipTraversalAutoImplOnly` bound while params without that get `V: TypeFoldable`.

Reply:
So this goes back to something we briefly touched on previously.

In this case (actually the only usage in the entire codebase) the field is just of type `V`, but in the general case if the field was of some other type generic over `V` should the constraint be on `V` or on that other generic type?  The current implementation does the latter, but I recall your view was that we should do the former.  Only if we do the former can we move the attribute to the type parameter instead.

Reply:
I would expect the constraint to still be on `V`, even if you have `GenericTy<V>` adding `V: SkipTraversalAutoImplOnly` should still be correct (and is how derives tend to work in general) :thinking:

Reply:
> even if you have `GenericTy<V>` adding `V: SkipTraversalAutoImplOnly` should still be correct

But then if some type `Foo<V, T>` contains a `ProjectionElem<V, T>`, we'll need to add a `#[skip_traversal]` attribute to Foo's `V` parameter even though that's an implementation detail of `ProjectionElem`, and indeed the attribute ends up propagating everywhere.  If instead we just bound Foo's field type, `ProjectionElem<V, T>: TypeFoldable` then it doesn't need to know anything about the internal implementation.

Reply:
I am confused. For `struct Foo<'tcx, T>(Field<'tcx, T>)`, does the derive emit the following?

```rust
impl<'tcx, T> TypeFoldable<TyCtxt<'tcx>> for Foo<'tcx, T>
where
    Field<'tcx, T>: TypeFoldable<'tcx>
{
    ...
}
```
because that seems like it should result in trait solver cycles as `TypeFoldable` is not coinductive. And if it emits `T: TypeFoldable<'tcx>` instead I don't get your comment.

If you have `struct Wrapper<'tcx, V>(ProjectionElem<V, Ty<'tcx>>)` I expect the derive to emit `V: TypeFoldable` no matter what you do in the impl for `ProjectionElem` so you kinda have to add `#[skip_traversal]` to the `V` of `Wrapper`.

I actually think that's kind of a good thing? Because it makes it clear that `Wrapper` also only expects to deal with trivially foldable types for `V`

Reply:
Yes, the derive currently emits what you have shown (with constraint on `Field`).  If it was instead `struct Foo<'tcx, T>(#[skip_traversal]` Field<'tcx, T>)` then instead of a `Field<'tcx, T>: TypeFoldable<'tcx>` the impl gets a `Field<'tcx, T>: SkipTraversalAutoImplOnly` constraint.

Maybe it's okay because it doesn't arise on any recursive type definitions?

Reply:
yeah, I am slightly in favor of moving it to the type parameter, maybe renaming it to `#[noop_traversable]` instead of `#[skip_traversal]` to signify that it's safe.

Not a strong requirement for merging this though